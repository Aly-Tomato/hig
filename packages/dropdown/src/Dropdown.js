import React, { Component } from "react";
import PropTypes from "prop-types";
import Downshift from "downshift";
import { TextFieldPresenter } from "@hig/text-field";
import Option from "./presenters/Option";
import composeEventHandlers from "./composeEventHandlers";

import "./dropdown.scss";

export default class Dropdown extends Component {
  static propTypes = {
    /**
     * Text describing what the field represents
     */
    label: PropTypes.string,
    /**
     * Instructional text for the field
     */
    instructions: PropTypes.string,
    /**
     * Placeholder text to render when an option has not been selected
     */
    placeholder: PropTypes.string,
    /**
     * Prevents user actions on the field
     */
    disabled: PropTypes.bool,
    /**
     * Text describing why the field is required
     */
    required: PropTypes.string,
    /**
     * An array of objects to choose from.
     */
    options: PropTypes.arrayOf(
      PropTypes.shape({
        label: PropTypes.string,
        value: PropTypes.string
      })
    ),
    /**
     * Called after user changes the value of the field
     */
    onChange: PropTypes.func,
    /**
     * Called when user moves focus from the field
     */
    onBlur: PropTypes.func,
    /**
     * Called when user puts focus onto the field
     */
    onFocus: PropTypes.func
  };

  render() {
    const {
      label,
      instructions,
      placeholder,
      required,
      disabled,
      onBlur,
      onFocus,
      options
    } = this.props;

    return (
      <Downshift itemToString={item => (item ? item.label : "")}>
        {({
          getInputProps,
          getItemProps,
          isOpen,
          selectedItem,
          toggleMenu
        }) => (
          <div className="hig__dropdown">
            <TextFieldPresenter
              {...getInputProps({
                label,
                instructions,
                required,
                placeholder,
                disabled,
                onBlur,
                onFocus: composeEventHandlers(toggleMenu, onFocus),
                readOnly: true // @TODO: toggle based on desired type of Dropdown
              })}
            />
            {isOpen && (
              <div className="hig__dropdown-v1__menu">
                {options.map((option, index) => (
                  <Option
                    {...getItemProps({
                      key: option.value,
                      index,
                      item: option,
                      selected: selectedItem === option
                    })}
                  >
                    {option.label}
                  </Option>
                ))}
              </div>
            )}
          </div>
        )}
      </Downshift>
    );
  }
}
