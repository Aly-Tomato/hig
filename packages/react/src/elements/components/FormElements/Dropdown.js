import React, { Component } from "react";
import PropTypes from "prop-types";

import DropdownAdapter from "../../../adapters/FormElements/DropdownAdapter";
import Option from "./Option";

export default class Dropdown extends Component {
  static defaultProps = {
    options: [],
    onChange: () => {}
  };

  constructor(props) {
    super(props);

    this.state = {
      open: false,
      value: this.getDefaultValue()
    };
  }

  getDefaultValue() {
    const { defaultValue, value } = this.props;

    if (value !== undefined) {
      return value;
    } else if (defaultValue !== undefined) {
      return defaultValue;
    }
    return "";
  }

  getRenderedValue() {
    const { value } = this.props;

    if (value !== undefined) {
      return value;
    }
    return this.state.value;
  }

  setSelectedValue = selectedOptionValue => {
    this.props.onChange(selectedOptionValue);

    this.setState({
      open: false,
      value: selectedOptionValue
    });
  };

  openDropdown = () => {
    this.setState({ open: true });
  };

  closeDropdown = () => {
    this.setState({ open: false });
  };

  render() {
    const selectedOption =
      this.props.options.find(o => o.value === this.getRenderedValue()) || {};

    return (
      <DropdownAdapter
        {...this.props}
        open={this.state.open}
        selectedOptionLabel={selectedOption.label}
        onTargetClick={this.openDropdown}
        onClickOutside={this.closeDropdown}
      >
        {this.props.options.map(option => (
          <Option
            key={option.value}
            {...option}
            selected={option.value === selectedOption.value}
            checked={option.value === selectedOption.value}
            onClick={this.setSelectedValue}
          />
        ))}
      </DropdownAdapter>
    );
  }
}

Dropdown.propTypes = {
  /**
   * {string} label for the the dropdown
   */
  label: PropTypes.string,
  /**
   * {string} instructions for the dropdown
   */
  instructions: PropTypes.string,
  /**
   * {string} placeholder for the dropdown
   */
  placeholder: PropTypes.string,
  /**
   * {bool} makes the dropdown disabled
   */
  disabled: PropTypes.bool,
  /**
   * {string} makes the field required
   */
  required: PropTypes.string,
  /**
   * {string} option that is selected on construction
   */
  value: PropTypes.string,
  /**
   * {string} default selected option
   */
  defaultValue: PropTypes.string,
  /**
   * {Array} array with objects, objects have a 'label' and a 'value'
   */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.string
    })
  ),
  /**
   * Calls the provided callback when option is changed
   */
  onChange: PropTypes.func,
  /**
   * Calls the provided callback when focus moves away from the dropdown
   */
  onBlur: PropTypes.func,
  /**
   * Calls the provided callback when the user focuses on the dropdown
   */
  onFocus: PropTypes.func,
  /**
   * Calls the provided callback when the user presses a key while the dropdown has focus
   */
  onKeypress: PropTypes.func
};
