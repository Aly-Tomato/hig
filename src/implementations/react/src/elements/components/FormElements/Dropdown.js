import React, { Component } from 'react';
import DropdownAdapter
  from '../../../adapters/FormElements/DropdownAdapter.js';
import * as PropTypes from 'prop-types';

const OptionAdapter = DropdownAdapter.Option;

class Dropdown extends Component {

  static propTypes = {
    label: PropTypes.string,
    instructions: PropTypes.string,
    placeholder: PropTypes.string,
    disabled: PropTypes.bool,
    required: PropTypes.string,
    value: PropTypes.string,
    defaultValue: PropTypes.string,

    options: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.string
    })),

    onChange: PropTypes.func,
    onBlur: PropTypes.func,
    onFocus: PropTypes.func,
    onKeypress: PropTypes.func
  }

  constructor(props) {
    super(props);

    var val = (this.props.value || this.props.defaultValue);
    var lab;
    
    if(this.props.options){
      this.props.options.forEach((v) => {
        if(v.value === val){
          lab = v.label;
        }
      }, this);
    }

    this.state = {
      open: false,
      value: val,
      selectedLabel: lab
    };

    
  }

  setSelectedValue = (selectedOption) => {
    if(selectedOption.value !== this.state.value){
      if(this.props.onChange){
        this.props.onChange(selectedOption);
      }
      this.setState({
        open: false,
        value: selectedOption.value,
        selectedLabel: selectedOption.label,
      });
    }
  };

  openDropdown = () => {
    this.setState({ open: true });
  };

  closeDropdown = () => {
    this.setState({ open: false });
  };

  render() {
    return (
      <DropdownAdapter
        label={this.props.label}
        instructions={this.props.instructions}
        placeholder={this.props.placeholder}
        disabled={this.props.disabled}
        required={this.props.required}
        
        open={this.state.open}
        selectedOptionLabel={this.state.selectedLabel}
        onTargetClick={this.openDropdown}
        onClickOutside={this.closeDropdown}

        onBlur={this.props.onBlur}
        onFocus={this.props.onFocus}
        onKeypress={this.props.onKeypress}
      >
        {this.props.options && this.props.options.map(option => (
          <OptionAdapter
            key={option.value}
            label={option.label}
            value={option.value}
            selected={option.value === this.state.value}
            onClick={this.setSelectedValue.bind(this, option)}
          />
        ))}
      </DropdownAdapter>
    );
  }
}


Dropdown.__docgenInfo = {
  props: {
    label: {
      description: "{string} label for the the dropdown"
    },
    instructions: {
      description: "{string} instructions for the dropdown"
    },
    placeholder: {
      description: "{string} placeholder for the dropdown"
    },
    disabled: {
      description: "{bool} makes the dropdown disabled"
    },
    required: {
      description: "{string} makes the field required"
    },
    value: {
      description: "{string} option that is selected on construction"
    },
    initialSelectedOption: {
      description: "{string} initial selected option"
    },
    options: {
      description: "{Array} array with objects, objects have a 'label' and a 'value'"
    },
    onChange: {
      description:
        "Calls the provided callback when option is changed"
    },
    onBlur: {
      description:
        "Calls the provided callback when focus moves away from the dropdown"
    },
    onKeypress: {
      description:
        "Calls the provided callback when the user presses a key while the dropdown has focus"
    },
    onFocus: {
      description:
        "Calls the provided callback when the user focuses on the dropdown"
    }
  }
};

export default Dropdown;
