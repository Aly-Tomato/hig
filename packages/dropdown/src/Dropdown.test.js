import renderer from "react-test-renderer";
import React from "react";

import Dropdown from "./Dropdown";

describe("Dropdown", () => {
  it("renders without props", () => {
    const tree = renderer.create(<Dropdown />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders with a label and an ID", () => {
    const tree = renderer
      .create(<Dropdown id="whoaBuddy" label="HIG Themes" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
