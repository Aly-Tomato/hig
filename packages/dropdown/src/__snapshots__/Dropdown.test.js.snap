// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders with a label and an ID 1`] = `
<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-labelledby="whoaBuddy-label"
  aria-owns={null}
  className="hig__dropdown"
  role="combobox"
>
  <div
    className="hig__dropdown__input-wrapper"
  >
    <div
      className="hig__text-field-v1 hig__text-field-v1--button"
    >
      <div
        className="hig__text-field-v1__content"
      >
        <div
          className="hig__text-field-v1__input-wrapper"
        >
          <span
            className="hig__text-field-v1__label-spacer"
          />
          <label
            className="hig__text-field-v1__label"
            htmlFor="whoaBuddy"
          >
            HIG Themes
          </label>
          <div
            className="hig__text-field-v1__input-row"
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-controls={null}
              aria-labelledby="whoaBuddy-label"
              autoComplete="nope"
              className="hig__text-field-v1__input hig__text-field-v1__input--no-value"
              disabled={undefined}
              id="whoaBuddy"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder={undefined}
              readOnly={true}
              type="button"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <span
      className="hig__dropdown__input-caret"
    >
      <div
        className="hig__icon hig__icon--24-size"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg width=\\"10\\" height=\\"6\\" viewBox=\\"0 0 10 6\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"caret-24\\"><polyline transform=\\"rotate(45 9.828 3)\\" points=\\"8 2 8 8 2 8\\" stroke=\\"#2E4258\\" stroke-width=\\"1.15\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/></svg>",
          }
        }
      />
    </span>
  </div>
</div>
`;

exports[`Dropdown renders without props 1`] = `
<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-labelledby="downshift-0-label"
  aria-owns={null}
  className="hig__dropdown"
  role="combobox"
>
  <div
    className="hig__dropdown__input-wrapper"
  >
    <div
      className="hig__text-field-v1 hig__text-field-v1--button"
    >
      <div
        className="hig__text-field-v1__content"
      >
        <div
          className="hig__text-field-v1__input-wrapper"
        >
          <div
            className="hig__text-field-v1__input-row"
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-controls={null}
              aria-labelledby="downshift-0-label"
              autoComplete="nope"
              className="hig__text-field-v1__input hig__text-field-v1__input--no-value"
              disabled={undefined}
              id="downshift-0"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder={undefined}
              readOnly={true}
              type="button"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <span
      className="hig__dropdown__input-caret"
    >
      <div
        className="hig__icon hig__icon--24-size"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg width=\\"10\\" height=\\"6\\" viewBox=\\"0 0 10 6\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"caret-24\\"><polyline transform=\\"rotate(45 9.828 3)\\" points=\\"8 2 8 8 2 8\\" stroke=\\"#2E4258\\" stroke-width=\\"1.15\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/></svg>",
          }
        }
      />
    </span>
  </div>
</div>
`;
